pkg/injector/injector.go

```diff
@@ -210,9 +210,19 @@ func (i *injector) handleRequest(w http.ResponseWriter, r *http.Request) {
 
 	ar := v1.AdmissionReview{}
 	_, gvk, err := i.deserializer.Decode(body, nil, &ar)
+	isDNSClusterFirst := true
 	if err != nil {
 		log.Errorf("Can't decode body: %v", err)
 	} else {
+		var pod corev1.Pod
+		if err = json.Unmarshal(ar.Request.Object.Raw, &pod); err != nil {
+			log.Warnf("could not unmarshal raw object: %v", err)
+		} else if pod.Spec.DNSPolicy != "" && pod.Spec.DNSPolicy != corev1.DNSClusterFirst {
+			log.Warnf("%q's DNSPolicy is not %q. Services may not be called", pod.Namespace+"/"+pod.Name, corev1.DNSClusterFirst)
+			isDNSClusterFirst = false
+		}
+	}
+	if isDNSClusterFirst {
```

Comment:
What's the purpose of checking if the value is true if it's hard-coded to true in line 213?

Reply:
Instead of skipping daprd injection when 'DNSPolicy' is specified with the value other than 'ClusterFirst', I suggest we should log warning message when it's configured to 'Default' but not stop the injection for daprd sidecar.

I examine [the doc](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-dns-config), and figure out 'ClusterFirstWithHostNet', 'ClusterFirst' (which is the default value), and 'None' should have no problem at all. Even for 'Default' policy, it is still arguable that user should make sure the host naming resolution is configured incorrectly, but I think it is helpful to log and doc this situation so that user can understand how to correct the configuration.

I suggest we make the following enhancement, based on the change you proposed:

```diff
        ar := v1.AdmissionReview{}
        _, gvk, err := i.deserializer.Decode(body, nil, &ar)
-       isDNSClusterFirst := true
        if err != nil {
                log.Errorf("Can't decode body: %v", err)
        } else {
                var pod corev1.Pod
                if err = json.Unmarshal(ar.Request.Object.Raw, &pod); err != nil {
                        log.Warnf("could not unmarshal raw object: %v", err)
-               } else if pod.Spec.DNSPolicy != "" && pod.Spec.DNSPolicy != corev1.DNSClusterFirst {
-                       log.Warnf("%q's DNSPolicy is not %q. Services may not be called", pod.Namespace+"/"+pod.Name, corev1.DNSClusterFirst)
-                       isDNSClusterFirst = false
+               } else if pod.Spec.DNSPolicy == corev1.DNSDefault {
+                       log.Warnf("%q's DNSPolicy is %q which may lead to daprd sidecar unable to talk to other K8S services.", pod.Namespace+"/"+pod.Name, pod.Spec.DNSPolicy)
                }
        }
-       if isDNSClusterFirst {
-               if !(utils.StringSliceContains(ar.Request.UserInfo.UID, i.authUIDs) || utils.StringSliceContains(systemGroup, ar.Request.UserInfo.Groups)) {
-                       log.Errorf("service account '%s' not on the list of allowed controller accounts", ar.Request.UserInfo.Username)
-               } else if ar.Request.Kind.Kind != "Pod" {
-                       log.Errorf("invalid kind for review: %s", ar.Kind)
-               } else {
-                       patchOps, err = i.getPodPatchOperations(&ar, i.config.Namespace, i.config.SidecarImage, i.config.SidecarImagePullPolicy, i.kubeClient, i.daprClient)
-                       if err == nil {
-                               patchedSuccessfully = true
-                       }
+
+       if !(utils.StringSliceContains(ar.Request.UserInfo.UID, i.authUIDs) || utils.StringSliceContains(systemGroup, ar.Request.UserInfo.Groups)) {
+               log.Errorf("service account '%s' not on the list of allowed controller accounts", ar.Request.UserInfo.Username)
+       } else if ar.Request.Kind.Kind != "Pod" {
+               log.Errorf("invalid kind for review: %s", ar.Kind)
+       } else {
+               patchOps, err = i.getPodPatchOperations(&ar, i.config.Namespace, i.config.SidecarImage, i.config.SidecarImagePullPolicy, i.kubeClient, i.daprClient)
+               if err == nil {
+                       patchedSuccessfully = true
                }
        }
```

Reply:
At the same time, on the daprd side, we also need to enhance the error message to hint users to diagnose the connection between sidecar and dapr-sidecar-injector.

```go
log.Errorf("error from operator stream: %s", err)
```